File: onyx-database/src/com/onyx/relationship/RelationshipReference.java
package com.onyx.relationship;

import com.onyx.exception.InvalidDataTypeForOperator;
import com.onyx.persistence.query.QueryCriteriaOperator;
import com.onyx.diskmap.serializer.ObjectBuffer;
import com.onyx.diskmap.serializer.ObjectSerializable;
import com.onyx.util.CompareUtil;

import java.io.IOException;

/**
 * Created by timothy.osborn on 3/19/15.
 *
 * Reference of a relationship
 */
public class RelationshipReference implements ObjectSerializable, Comparable
{

    @SuppressWarnings("unused")
    public RelationshipReference()
    {

    }

    public RelationshipReference(Object identifier, long partitionId)
    {
        this.identifier = identifier;
        this.partitionId = partitionId;
    }

    public Object identifier;
    public long partitionId;

    @Override
    public int hashCode()
    {
        if(identifier == null)
        {
            return super.hashCode();
        }
        return identifier.hashCode() + new Long(partitionId).hashCode();
    }

    @Override
    public boolean equals(Object obj)
    {
        if(obj instanceof RelationshipReference)
        {
            final RelationshipReference comp = (RelationshipReference) obj;
            if(comp.identifier.getClass() != identifier.getClass())
                return false;
            try
            {
                return (comp.partitionId == partitionId && CompareUtil.compare(comp.identifier, identifier, QueryCriteriaOperator.EQUAL));
            }catch (InvalidDataTypeForOperator e)
            {
                return false;
            }
        }
        return false;
    }

    @Override
    public void writeObject(ObjectBuffer buffer) throws IOException
    {
        buffer.writeLong(partitionId);
        buffer.writeObject(identifier);
    }

    @Override
    public void readObject(ObjectBuffer buffer) throws IOException
    {
        partitionId = buffer.readLong();
        identifier = buffer.readObject();
    }

    @Override
    public String toString()
    {
        return "Identifier " + identifier.toString() + " Partition ID " + partitionId;
    }

    @Override
    @SuppressWarnings("unchecked")
    public int compareTo(Object o) {

        if(o instanceof RelationshipReference)
        {
            if(this.partitionId < ((RelationshipReference) o).partitionId)
                return -1;
            else if(this.partitionId > ((RelationshipReference) o).partitionId)
                return 1;

            if(((RelationshipReference) o).identifier.getClass() == this.identifier.getClass()
                && this.identifier instanceof Comparable)
                return ((Comparable) this.identifier).compareTo(((RelationshipReference) o).identifier);

            if(this.hashCode() < o.hashCode())
                return -1;
            else if(this.hashCode() > o.hashCode())
                return 1;
            else
                return 0;
        }
        
        return new Integer(this.hashCode()).compareTo(o.hashCode());
    }
}

File: onyx-database/src/main/kotlin/com/onyx/interactors/relationship/data/RelationshipReference.kt
package com.onyx.interactors.relationship.data

import com.onyx.buffer.BufferStream
import com.onyx.buffer.BufferStreamable
import com.onyx.exception.BufferingException
import com.onyx.persistence.context.SchemaContext

/**
 * Created by timothy.osborn on 3/19/15.
 *
 * Reference of a relationship
 */
class RelationshipReference @JvmOverloads constructor(var identifier: Any? = "", var partitionId: Long = 0L) : BufferStreamable, Comparable<RelationshipReference> {

    override fun read(buffer: BufferStream) {
        partitionId = buffer.long
        identifier = buffer.value
    }

    override fun write(buffer: BufferStream) {
        buffer.putLong(partitionId)
        buffer.putObject(identifier)
    }

    @Throws(BufferingException::class)
    override fun read(buffer: BufferStream, context: SchemaContext?) {
        this.read(buffer)
    }

    @Throws(BufferingException::class)
    override fun write(buffer: BufferStream, context: SchemaContext?) {
        this.write(buffer)
    }

    override fun compareTo(other: RelationshipReference): Int = when {
        this.partitionId < other.partitionId -> -1
        this.partitionId > other.partitionId -> 1
        other.identifier!!.javaClass == this.identifier!!.javaClass && this.identifier is Comparable<*> -> @Suppress("UNCHECKED_CAST") (this.identifier as Comparable<Any>).compareTo(other.identifier!! as Comparable<Any>)
        else -> 0
    }

    override fun hashCode(): Int = when (identifier) {
        null -> partitionId.hashCode()
        else -> identifier!!.hashCode() + partitionId.hashCode()
    }

    override fun equals(other:Any?):Boolean = other != null && (other as RelationshipReference).partitionId == partitionId && other.identifier == identifier
}